text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,138 @@
+# Meta
+[meta]: #meta
+- Name: Buildpacks should be able to define the default process type for an app
+- Start Date: 8/20/20
+- Author(s): natalieparellano
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: n/a
+
+# Summary
+[summary]: #summary
+
+Today, if a buildpacks user would like to define the default process type for their app, they need to pass `-process-type` to the exporter. This is cumbersome because a user might not know which process types are available until after the build has completed. Additionally, [this spec PR](https://github.com/buildpacks/spec/pull/137) would force some users who are currently relying on the lifecycle to specify the default process type to pass `-process-type` instead. This RFC proposes that buildpacks should be able to define the default process type for an app, while retaining the ability for a user to specify the desired default.
+
+# Definitions
+[definitions]: #definitions
+
+launcher - a [lifecycle binary](https://github.com/buildpacks/spec/blob/main/platform.md#launcher) that is responsible for starting app processes. An app image built with CNBs will have the launcher's path (or a symlink to it) as its [entrypoint](https://github.com/opencontainers/image-spec/blob/master/config.md#properties).
+
+process type - a CNB construct representing a process that can be started by the launcher. A process type has the following properties as described [here](https://github.com/buildpacks/spec/blob/main/buildpack.md#launchtoml-toml): type (string), command (string), arguments (array of string), direct (boolean). Buildpacks declare process types during the `build` phase by writing entries into `<layers>/launch.toml`.
+
+default process type - describes the process type that will be started by the launcher when the app image is run. It can be overriden by overriding the image entrypoint. Example: an image with default process type of web would have an entrypoint of `/cnb/process/web` which is a symlink that points to `/cnb/lifecycle/launcher`. The launcher uses argv0 to determine which process to start.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this? Easier and more intuitive for end users.
+- What use cases does it support? Apps with one or more process type(s).
+- What is the expected outcome? The experience should be seamless for end users. Buildpacks should produce a sensible default process type without the user having to know or care that process types exist. The platform should still be able to override the default process type by passing `-process-type` to the lifecycle.
```

Comment:
If one day in the mid-future `pack` no longer passes `web`, the *hope* is that the buildpacks will produce a runnable image with `web` because the buildpacks were updated. But in truth, this will require buildpack authors to make a change before this future date.

Is there some world in which we might use the latest buildpack API to determine if `pack` should send `web` in? If all your buildpacks are on the old buildpack API, `pack` will always send in `web`. If any of your buildpacks are on this new spec version, `pack` no longer sends `web`. 

Maybe we could go further and perhaps make `default = true` for buildpacks that produce `web` processes that are running on the old buildpack spec. 

I'm still a bit worried about buildpacks not being updated and an impending deadline to update them that the authors may not be aware of.

Reply:
> Maybe we could go further and perhaps make default = true for buildpacks that produce web processes that are running on the old buildpack spec.

This seems like a clean way of solving the problem, obviating the need for `pack` to send `web` at all. It is pushing knowledge about `web` back into the lifecycle, but only for older buildpack APIs (and defaulting to `web` is already the behavior for older platform APIs).

Reply:
> Maybe we could go further and perhaps make default = true for buildpacks that produce web processes that are running on the old buildpack spec.

Practically I think this ^ suggestion would produce behavior that is identical to this one [from the other thread](https://github.com/buildpacks/rfcs/pull/110#discussion_r504918609)

> Have lifecycle default to web when no default process is set. 

The latter framing seems easier to explain to me, because we aren't creating new rules about old APIs.

Reply:
Can we capture this migration plan in this RFC?

Do we need a way for the platform to pass a default process type that is separate from the users or is that just a platform feature ala `pack`?

Reply:
> Maybe we could go further and perhaps make default = true for buildpacks that produce web processes that are running on the old buildpack spec.

After another day of thinking about this I have decided that @jabrown85 is correct here and I was wrong above. Tying this to the buildpack API allows us to phase out this behavior as buildpacks upgrade.

Reply:
Added a section to describe the migration approaches discussed: https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR139-R160

Reply:
Migration plan covered here: https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR103-R114

Reply:
Migration plan covered here: https://github.com/buildpacks/rfcs/pull/110/files#diff-00a2fb620563b974f876628cc19e746a4a96606f9dea0fd58dc1bb13f010b4abR103-R114